lib/dotcom/utils/service_date_time.ex

```diff
@@ -0,0 +1,228 @@
+
+  However, the `service_range/1` function *is* exclusive.
+  Given a date_time in today's service range, it will return :today.
+  """
+
+  require Logger
+
+  use Timex
+
+  import Dotcom.Utils.DateTime, only: [coerce_ambiguous_time: 1, now: 0, timezone: 0]
+
+  alias Dotcom.Utils
+
+  @type service_range() :: :past | :today | :this_week | :next_week | :later
+
+  @service_rollover_time Application.compile_env!(:dotcom, :service_rollover_time)
+  @timezone timezone()
+
+  @doc """
+  Returns the time at which service rolls over from 'today' to 'tomorrow'.
+  """
+  def service_rollover_time(), do: @service_rollover_time
+
+  @doc """
+  Get the service date for the given date_time.
+  If the time is before 3am, we consider it to be the previous day.
+  """
+  @spec service_date() :: Date.t()
+  @spec service_date(DateTime.t()) :: Date.t()
+  def service_date(date_time \\ now()) do
+    if date_time.hour < @service_rollover_time.hour do
+      Timex.shift(date_time, hours: -@service_rollover_time.hour)
+      |> coerce_ambiguous_time()
+      |> Timex.to_date()
+    else
+      Timex.to_date(date_time)
+    end
+  end
+
+  @doc """
+  The most specific service range for the given date_time.
+  A date_time for today can be in the range for today and this week.
+  But, we just return :today since that is the most specific range.
+  """
+  @spec service_range(DateTime.t()) :: service_range()
+  def service_range(date_time) do
+    Enum.find(
+      [
+        &service_past?/1,
+        &service_today?/1,
+        &service_this_week?/1,
+        &service_next_week?/1,
+        &service_later?/1
+      ],
+      fn f ->
+        f.(date_time)
+      end
+    )
+    |> Kernel.inspect()
+    |> Kernel.then(fn module -> Regex.run(~r/_(\w+)\?/, module) end)
+    |> List.last()
+    |> String.to_atom()
+  end
+
+  @doc """
+  Get the beginning of the service day for the day after the given date_time.
+  """
+  @spec beginning_of_next_service_day() :: DateTime.t()
+  @spec beginning_of_next_service_day(DateTime.t()) :: DateTime.t()
+  def beginning_of_next_service_day(datetime \\ now()) do
+    datetime
+    |> end_of_service_day()
+    |> Timex.shift(microseconds: 1)
+    |> coerce_ambiguous_time()
+  end
+
+  @doc """
+  Get the beginning of the service day for the given date_time.
+  """
+  @spec beginning_of_service_day() :: DateTime.t()
+  @spec beginning_of_service_day(DateTime.t()) :: DateTime.t()
+  def beginning_of_service_day(date_time \\ now()) do
+    date_time
+    |> service_date()
+    |> Timex.to_datetime(@timezone)
+    |> coerce_ambiguous_time()
+    |> Map.put(:hour, @service_rollover_time.hour)
+  end
+
+  @doc """
+  Get the end of the service day for the given date_time.
+  """
+  @spec end_of_service_day() :: DateTime.t()
+  @spec end_of_service_day(DateTime.t()) :: DateTime.t()
+  def end_of_service_day(date_time \\ now()) do
+    date_time
+    |> service_date()
+    |> Timex.to_datetime(@timezone)
+    |> coerce_ambiguous_time()
```

Comment:
Curiosity: Is this first `coerce_ambiguous_time()` necessary, given that there's another one two lines down?

(All of the tests pass even if both `coerce_ambiguous_time()` calls are removed, so I'm actually not sure whether it's that neither one is necessary, or that there's some missing test coverage?)

Reply:
Yes, they are both necessary because `to_datetime/2` can return an ambiguous date_time and `shift` does not take an ambiguous date_time. You can be ambiguous both times. Anything that can return a date_time within a daylight shift window has to be disambiguated.

Reply:
Gotcha - thanks!

Can you add a test that reliably fails for that reason? Otherwise we'll end up refactoring, losing one of the `coerce_ambiguous_time()` calls, and then eventually discarding the existing test as flaky if it fails one in 150 times (or one in ~4000 times).

Reply:
Not really because we don't have a mock for Timex. We would need to somehow have `to_datetime` return an `{:error, term()}` then also have `shift` return an `{:error, term()}` and that probably isn't ever going to happen given we're feeding it a date. I really think we have this base covered.

Reply:
Wait a minute... this function can take a `date_time` as an argument, so it doesn't need to mock Timex. Just pass in that one time in March (or November?) where the first coercion matters, and then in a different test, pass in that one time an hour later where the other one matters.